core/src/nodes/protocols_handler.rs

```diff
@@ -0,0 +1,628 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::prelude::*;
+use nodes::handled_node::{NodeHandler, NodeHandlerEndpoint, NodeHandlerEvent};
+use std::{io, marker::PhantomData, time::Duration};
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_timer::Timeout;
+use upgrade::{self, apply::UpgradeApplyFuture, DeniedConnectionUpgrade};
+use void::Void;
+use {ConnectionUpgrade, Endpoint};
+
+/// Handler for a set of protocols for a specific connection with a remote.
+///
+/// This trait should be implemented on struct that hold the state for a specific protocol
+/// behaviour with a specific remote.
+///
+/// # Handling a protocol
+///
+/// Protocols with the remote can be opened in two different ways:
+///
+/// - Dialing, which is a voluntary process. In order to do so, make `poll()` return a
```

Comment:
```suggestion
/// - Dialing, which is a voluntary process. In order to do so, make `poll()` return an
```
Use "an" before words beginning with vowel sounds, otherwise use "a".

Reply:
In my mind the word it applies to is "variant" and not "outbound". But I don't know whether that's correct.

Reply:
The article matches with the noun, but it applies to the word immediately after it, e.g. a big octopus, an extraordinary dog. In your case `OutboundSubstreamRequest` variant is more like a two word noun, like cantilever bridge, etc.